### DISCUSSION TITLE
What to do when fragmenter fails?

### DISCUSSION THREAD

**Discussion by pavankum** [openforcefield/openff-bespokefit#178] [Q&A] [answered]
Hi, I tried one of the heterocycles from vehicle set and it fails fragmentation step. Do you think we can skip fragmentation step for those cases and generate torsion scans for all the torsions present, what would be the downside of doing that? Here is the smiles string `'[H:15][C:1]1=[N:6][N@@:9]2[N:7]([C:3](=[O:11])[C:2](=[O:10])[C:4](=[O:12])[N:8]2[O:14][C:5]1=[O:13])[H:16]'`.

**Comment by jthorton** [comment, marked as answer]
Hi @pavankum this is a bit of a strange one by default we tell BespokeFit to ignore these types of torsions as they are considered non-rotatable and probably well described by the base force field, the default SMIRKS pattern which controls which torsions we try to optimise can be found [here](https://github.com/openforcefield/openff-bespokefit/blob/e4ea0e66e5fb46d1df3d420143cc3924cf8a50b6/openff/bespokefit/workflows/bespoke.py#L68) and can be changed for something more general to make BespokeFit target these in-ring torsions. This will then require users to skip the fragmentation step as you point out to do this in the workflow simply set the [fragmentation_engine](https://github.com/openforcefield/openff-bespokefit/blob/e4ea0e66e5fb46d1df3d420143cc3924cf8a50b6/openff/bespokefit/workflows/bespoke.py#L119) to `None` while this should work there might be bugs along the way so I would be careful using BespokeFit for ring torsions.

**Comment by j-wags** [comment]
Yeah - I dug into this today and I think the current behavior is correct. The input molecule in question has no rotatable bonds (I think) and looks like this:


<img width="260" alt="Screen Shot 2022-05-26 at 8 52 22 AM" src="https://user-images.githubusercontent.com/16329175/170525396-d420f60f-7583-489e-beae-1a1f1e6ca6b3.png">

I think acceptable behaviors from bespokefit here would be either outputting a functionally empty/unchanged OFFXML, or raising a descriptive error. The current code raises a descriptive error, so I think this is fine. 

A minimal reproducing example is:


```
openff-bespoke executor run --smiles "[H:15][C:1]1=[N:6][N@@:9]2[N:7]([C:3](=[O:11])[C:2](=[O:10])[C:4](=[O:12])[N:8]2[O:14][C:5]1=[O:13])[H:16]"  --workflow  "default" --output "acetaminophen.json" --output-force-field "acetaminophen.offxml" --n-qc-compute-workers 2 --qc-compute-n-cores   1 --default-qc-spec xtb gfn2xtb none
```

and it yields:

```
[✓] bespoke executor launched

1. preparing the bespoke workflow                                                                                       

[✓] 1 molecules found
[✓] fitting schemas generated

2. submitting the workflow                                                                                              

[✓] the following workflows were submitted
┏━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━┳━━━━━━┓
┃ ID ┃ SMILES                            ┃ NAME ┃ FILE ┃
┡━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━╇━━━━━━┩
│ 1  │ C1=N[N@@]2NC(=O)C(=O)C(=O)N2OC1=O │      │      │
└────┴───────────────────────────────────┴──────┴──────┘

3. running the fitting pipeline                                                                                         

[x] fragmentation failed

 "No fragments could be generated for the parent molecule. This likely means that the bespoke parameters that you have  
 generated and are trying to fit are invalid. Please raise an issue on the GitHub issue tracker for further assistance."

outputs have been saved to acetaminophen.json                                                                                                                    
```

**Comment by pavankum** [comment]
Thank you @jthorton & @j-wags!!
